src/nbd.c

```diff
@@ -0,0 +1,1286 @@
+	nbd_dev->index = 0;
+
+	/* maybe reuse the socket to get final statistics/error message? */
+	g_close(nbd_dev->sock, NULL);
+	nbd_dev->sock = -1;
+	g_clear_pointer(&nbd_dev->dev, g_free);
+
+	res = TRUE;
+	goto out;
+
+nla_put_failure:
+	res = FALSE;
+	g_set_error(error, G_IO_ERROR, G_IO_ERROR_FAILED, "netlink NLA_PUT failed");
+
+out:
+	if (nl) {
+		nl_close(nl);
+		nl_socket_free(nl);
+	}
+	return res;
+}
+
+struct RaucNBDContext {
+	gint sock;
+
+	/* configuration */
+	guint64 data_size;
+	gchar *url;
+	gchar *tls_cert; /* local file or PKCS#11 URI */
+	gchar *tls_key; /* local file or PKCS#11 URI */
+	gchar *tls_ca; /* local file */
+	gboolean tls_no_verify;
+	GStrv headers; /* array of strings such as 'Foo: bar' */
+
+	/* runtime state */
+	CURLM *multi;
+	gboolean done;
+	struct curl_slist *headers_slist;
+
+	/* statistics */
+	struct RaucStats dl_size, dl_speed,
+	                 namelookup, connect, starttransfer, total;
+	/* FIXME add block histogram */
+};
+
+struct RaucNBDTransfer {
+	struct RaucNBDContext *ctx;
+
+	/* curl */
+	CURL *easy;
+	char errbuf[CURL_ERROR_SIZE];
+
+	struct nbd_request request;
+	struct nbd_reply reply;
+	gboolean done;
+	guint errors;
+
+	guint8 *buffer;
+	curl_off_t buffer_size;
+	curl_off_t buffer_pos;
+
+	/* configure request */
+	guint64 content_size;
+	guint64 current_time; /* date header from server */
+	guint64 modified_time; /* last-modified header from server */
+};
+
+static size_t write_cb(char *ptr, size_t size, size_t nmemb, void *userdata)
+{
+	struct RaucNBDTransfer *xfer = userdata;
+	size_t remaining = 0;
+
+	g_assert_cmpint(size, ==, 1); /* according to the docs, size is always 1 */
+
+	remaining = xfer->buffer_size - xfer->buffer_pos;
+	if (remaining < nmemb) {
+		return 0;
+	}
+	memcpy(xfer->buffer + xfer->buffer_pos, ptr, nmemb);
+	xfer->buffer_pos += nmemb;
+
+	return nmemb;
+}
+
+static size_t header_cb(char *buffer, size_t size, size_t nitems, void *userdata)
+{
+	struct RaucNBDTransfer *xfer = userdata;
+	g_autofree gchar *header = NULL;
+	g_auto(GStrv) h_pair = NULL;
+
+	/* TODO read data and last modified? */
+
+	g_assert_cmpint(size, ==, 1); /* according to the docs, size is always 1 */
+
+	/* make sure we have a 0-terminated lowercase string */
+	header = g_strchomp(g_ascii_strdown(buffer, nitems));
+
+	h_pair = g_strsplit(header, ": ", 2);
+	if (g_strv_length(h_pair) < 2)
+		return nitems;
+
+	if (g_str_equal(h_pair[0], "content-range")) {
+		g_auto(GStrv) h_elements = NULL;
+		g_auto(GStrv) h_range = NULL;
+		gchar *endptr = NULL;
+		guint64 range_size = 0;
+
+		h_elements = g_strsplit(h_pair[1], " ", 2);
+		g_assert(g_strv_length(h_elements) == 2);
+
+		h_range = g_strsplit(h_elements[1], "/", 2);
+		g_assert(g_strv_length(h_range) == 2);
+
+		g_assert_cmpstr(h_range[0], ==, "0-3");
+		g_assert(strlen(h_range[1]) > 0);
+		g_assert_cmpstr(h_range[1], !=, "*");
+
+		errno = 0;
+		range_size = g_ascii_strtoull(h_range[1], &endptr, 10);
+		g_assert(errno == 0);
```

Comment:
Given that g_assert can be disabled, should it be used for all kinds of error checking outside of tests?

Reply:
Yes, this needs better error handling.